File: src/main/java/site/gachontable/gachontablebe/domain/admin/usecase/ReadyUser.java
package site.gachontable.gachontablebe.domain.admin.usecase;

import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import site.gachontable.gachontablebe.domain.pub.domain.Pub;
import site.gachontable.gachontablebe.domain.waiting.domain.Waiting;
import site.gachontable.gachontablebe.domain.waiting.domain.repository.WaitingRepository;
import site.gachontable.gachontablebe.domain.waiting.type.Status;
import site.gachontable.gachontablebe.global.biztalk.SendBiztalk;

import java.util.HashMap;
import java.util.List;

@Service
@RequiredArgsConstructor
public class ReadyUser {

    private final WaitingRepository waitingRepository;
    private final SendBiztalk sendBiztalk;

    @Value("${biztalk.templateId.ready}")
    private String TEMPLATE_CODE;

    public void execute(Pub pub) {
        List<Waiting> waitings = waitingRepository
                .findAllByPubAndWaitingStatusOrWaitingStatusOrderByCreatedAtAsc(pub, Status.WAITING, Status.AVAILABLE);

        if (waitings.size() < 3) {
            return;
        }

        Waiting waiting = waitings.get(2);

        HashMap<String, String> variables = new HashMap<>();
        variables.put("#{pub}", pub.getPubName());
        variables.put("#{username}", waiting.getUser().getUsername());
        variables.put("#{callNumber}", waiting.getTel().substring(9));
        variables.put("#{waitingId}", waiting.getWaitingId().toString());

        sendBiztalk.execute(TEMPLATE_CODE, waiting.getTel(), variables);
    }
}

File: src/main/kotlin/site/gachontable/domain/admin/service/ReadyUser.kt
package site.gachontable.domain.admin.service

import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import site.gachontable.domain.pub.domain.Pub
import site.gachontable.domain.waiting.domain.Waiting
import site.gachontable.domain.waiting.port.out.WaitingRepository
import site.gachontable.domain.waiting.type.Status
import site.gachontable.infra.biztalk.SendBiztalk
import java.util.*

@Service
class ReadyUser(
    private val waitingRepository: WaitingRepository,
    private val sendBiztalk: SendBiztalk,

    @Value("\${biztalk.templateId.ready}")
    private val templateCode: String,
) {
    fun execute(pub: Pub) {
        val waitings: MutableList<Waiting> = waitingRepository.findAllByPubAndWaitingStatusInOrderByCreatedAtAsc(
            pub, mutableListOf<Status>(Status.WAITING, Status.AVAILABLE)
        )

        if (waitings.size < 3) {
            return
        }

        val waiting: Waiting = waitings[2]

        val variables = HashMap<String, String>()
        variables.put("#{pub}", pub.pubName)
        variables.put("#{username}", waiting.user.username)
        variables.put("#{waitingId}", waiting.id.toString())

        sendBiztalk.execute(templateCode, waiting.tel, variables)
    }
}
